File: crates/qtbridge-runtime/src/cpp/rustobjectgetter.cpp
// Copyright (C) 2026 The Qt Company Ltd.
// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only
#include "rustobjectgetter.h"
#include <QObject>

const uint8_t* getRustObjectRcPtr(const QObject& qobj)
{
    if (auto* getter = dynamic_cast<const RustObjectGetter*>(&qobj))
        return static_cast<const uint8_t*>(getter->getRustObjectRcPtr());
    return nullptr;
}

const uint8_t* getRustProxy(const QObject& qobj)
{
    if (auto* getter = dynamic_cast<const RustObjectGetter*>(&qobj))
        return static_cast<const uint8_t*>(getter->getRustProxy());
    return nullptr;
}

File: crates/qtbridge-runtime/src/rustobjectgetter.rs
// Copyright (C) 2026 The Qt Company Ltd.
// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only

#[cxx::bridge]
pub mod ffi {
    unsafe extern "C++" {
        include!("qtbridge-type-lib/src/generated/core/qobject/cpp/qobject.h");
        type QObject = qtbridge_type_lib::QObject;
    }

    unsafe extern "C++" {
        include!("cpp/rustobjectgetter.h");

        #[rust_name = "get_rust_proxy"]
        fn getRustProxy(qobj: &QObject) -> *const u8;
    }
}

pub use ffi::get_rust_proxy;
